### PR TITLE
Abstract data provider out

### PR DESCRIPTION
**Pull request by e0ipso**
Introducing the concept of `RestfulDataProviderInterface`. That will allow us to have a data provider for SearchAPI, another for QueryDBTNG, …

### CODE REVIEW

**Comment by e0ipso** [comment]
@amitaibu green tests. In the next PR I will add an example of a fully working resource for the variable table.

**Comment by amitaibu** [comment]
Wow, big PR :)

**Comment by e0ipso** [comment]
Nah, it's mostly moving stuff around

**Comment by amitaibu** [comment]
> In the next PR I will add an example of a fully working resource for the variable table.

I'd actually be happy to see that PR before this goes in, just so it would be easier for me to wrap my head around this. Would that be ok?

**Comment by amitaibu** [comment]
(I understand what you do here, I just want to see what it takes for an implementing module to create a new data provider)

File: plugins/restful/RestfulBase.php [hunk 1]
```diff
@@ -141,41 +170,30 @@ public static function cleanRequest(&$request) {
    *   route pattern.
    */
   public static function controllersInfo() {
+    // Provide sensible defaults for the HTTP methods. These methods (index,
```
Comment (amitaibu) [inline plugins/restful/RestfulBase.php:173]: I'm not sure "index" really better than "list"?

Comment (e0ipso) [inline plugins/restful/RestfulBase.php:173]: I agree 100%. Actually I did use _list_, but it turns out that you cannot use reserved words like _list_ or _delete_ as your method name. I belive that _index_ and _remove_ are close seconds.

**Comment by e0ipso** [comment]
Since there are no broken tests and no big comments in here, can we merge this and fix whatever in #189?

**Comment by amitaibu** [comment]
> no big comments in here

Actually, that's because I've added all the comments in #189. I would prefer to close this one in favor of #188 - as the renames are an important part of the fixes.

File: plugins/restful/RestfulBase.php (hunk 1 above)
Comment (amitaibu) [inline plugins/restful/RestfulBase.php:173]: Are you sure about `delete`, I think it is not reserved.

**Comment by e0ipso** [comment]
Closed in favor of #189.

File: plugins/restful/RestfulBase.php (hunk 1 above)
Comment (amitaibu) [inline plugins/restful/RestfulBase.php:173]: @mateu-aguilo-bosch ^^

**Pull request closed without merging**